### PR TITLE
Darkmode toggle take 2

### PR DESCRIPTION
**Pull request by JuanitoFatas**
## Summary

This Pull Request adds a dark mode switch (html button) in the top right nav menu. The switch defaults to user‘s system darkmode setting. Once toggled will store preference in the `Window.localStorage`.  The switch icons are `☼, U+263C` and `☾, U+263E`.

The switch is managed by a Stimulus controller `theme-switch_controller.js`. When the page loads, we check the OS and localStorage to find user‘s preference and update the switch css class. When user clicks it, we toggle the css class and store preference in localStorage.

When we use [tailwindcss-theming] plugin, the tailwind‘s original colors are all gone. So I added back what we are using in `theme.config.js`, then extract all places using colors in the view into CSS classes.

Added css class for development search query‘s code block display.

Implements #290.

[tailwindcss-theming]: https://github.com/hawezo/tailwindcss-theming

## Try at local

```zsh
$ git clone rubyapi/rubyapi && cd rubyapi
$ git co -b test-pr-323 && git am -3 https://github.com/rubyapi/rubyapi/pull/323
$ bundle install && yarn install && bin/rails s
$ open http://localhost:3000
```

## Review

We could use the css classes separately or extract the common values into generic names. Suggestions welcome.

## Screenshots 

| mobile light | mobile dark |
| ----------- | ------------- |
| スクリーンショット 2020-03-15 23 45 49  | スクリーンショット 2020-03-15 23 45 54 |

### CODE REVIEW

**Comment by AlexWayfer** [comment]
Looks… good to me, I think. Just a small notice about icons: the Moon is filled (white), but the Sun is outlined (transparent). 

Let's look what icons are used… oh, Unicode symbols.

I suggest black one, not white: https://unicode-table.com/en/2600/

But it can be rendered colorful, like Emoji, and probably we should use some (SVG) monotone external icons.

Also I've faced recently with problems of star icon ([this one](https://unicode-table.com/en/1F7CA/)) on macOS and iOS, so we should check Unicode symbols there.

File: app/javascript/packs/app/application.scss [hunk 1]
```diff
@@ -71,4 +71,18 @@ pre:not(.ruby) {
...
+.dark-switch:before {
+  content: '☼';
+}
+
+.dev-search-query {
+  @apply bg-gray-900 border-gray-800;
+  color: #5FB3B3;
+}
+
+/* purgecss end ignore */
```
Comment (colby-swandale) [inline app/javascript/packs/app/application.scss]: It this a duplicate?

File: app/javascript/packs/app/application.scss [hunk 2]
```diff
@@ -71,4 +71,18 @@ pre:not(.ruby) {
   }
 }
 
+.light-switch:before {
+  content: '☾';
```
Comment (colby-swandale) [inline app/javascript/packs/app/application.scss]: I've been using https://fontawesome.com for the UI icons, which you can find [here](https://github.com/rubyapi/rubyapi/blob/master/app/javascript/packs/app/icons.js). Would we be able to use the icons here as well?

Comment (JuanitoFatas) [reply app/javascript/packs/app/application.scss]: Let‘s do this in later Pull Request? I did not find ones on fontawesome that are better than current unicode ones. Ideally we could have CSS icons like the ones in https://justjavascript.com/.

Comment (AlexWayfer) [reply app/javascript/packs/app/application.scss]: > I did not find ones on fontawesome that are better than current unicode ones.

Moon: https://fontawesome.com/icons/moon?style=solid

Sun: https://fontawesome.com/icons/sun?style=solid (a bit another, but seems OK)

Bulb icons (like in Twitter) are paid in Font Awesome.

Also sometimes I think we can have too many icons/buttons in the header, and there will be some drop-down menu, and then it'd be better to do just toggle with text, without icon.

**Comment by AlexWayfer** [comment]
@JuanitoFatas there are conflicts. Can you please resolve them? And do you need a help? I'm ready to try. :heart:

**Comment by JuanitoFatas** [comment]
@AlexWayfer Conflicts resolved. Feel free to send a PR to this branch.

**Comment by AlexWayfer** [comment]
> @AlexWayfer Conflicts resolved. Feel free to send a PR to this branch.

Done in #357.

**Comment by colby-swandale** [comment]
Apologies for the delay again 🙇‍♂️. I am looking at this now.

**Comment by AlexWayfer** [comment]
> Apologies for the delay again 🙇‍♂️. I am looking at this now.

No problem. I just tried to simplify selectors and actions in #357, and waiting when @JuanitoFatas will look at this.

**Comment by AlexWayfer** [comment]
@JuanitoFatas can you please rebase this PR? Or should we a new one with such changes?

**Comment by colby-swandale** [comment]
Ah i have a couple of big changes to this PR that i will try to push in the next few days. I'm happy to do the rebase myself

**Comment by colby-swandale** [comment]
👋  @JuanitoFatas I'm really really sorry for the super long delay in getting this merged. I had a close look at the tailwindcss theming plugin and ended up not liking having to specify colours in a separate file. I really like specifying the colours inline via the tailwindcss colours, ie: `text-gray-600 dark:text-gray-200` etc.

Your stimulus component was great and really liked the approach! But I ended up using the the awesome-font icon from #357 to keep the UI consistent. 

The header is a bit of a mess at the moment, so I've hidden the theme icon on every page except the homepage for mobile devices. I'm about 1/2 way through redesigning the header from scratch and adding a hamburger menu option that will look much better and fit everything more nicely!

Thanks so much again for working on this!

**Pull request merged**